src/id_allocator.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::result;
+
+// Error type for IdAllocator usage.
+#[derive(Debug)]
+pub enum Error {
+    // All ids from the range specified are exhausted.
+    Overflow,
+    // An id that is not part of the specified range was requested to be released.
+    InvalidInput(u32),
+    // There are no more IDs available in the manage range
+    ResourceExhausted,
+}
+
+pub type Result = result::Result<u32, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Overflow => write!(f, "Id counter overflowed."),
+            InvalidInput(id) => write!(f, "Specified id: {} is not in the range.", id),
+            ResourceExhausted => {
+                write!(f, "There are no more available ids in the specified range.")
+            }
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # use vm_allocator::IdAllocator;
+///
+/// let mut legacy_irq_allocator = IdAllocator::new(5, 23);
+///
+/// for _ in 1..10 {
+///     let _id = legacy_irq_allocator.allocate_id().unwrap();
+/// }
+/// let irq = 8;
+/// let freed_id_result = legacy_irq_allocator.free_id(irq);
+/// if freed_id_result.is_err() {
+///     println!("Failed to free the specified id: {:?}", freed_id_result);
+/// }
+/// ```
+
+/// Id allocator representation.
+pub struct IdAllocator {
+    // Begining of the range of IDs that we want to manage.
+    range_base: u32,
+    // First available id that was never allocated.
+    pub next_id: u32,
+    // End of the range of IDs that we want to manage.
+    range_end: u32,
+    // Set of all freed ids that can be reused at subsequent allocations.
+    pub freed_ids: BTreeSet<u32>,
+}
+
+impl IdAllocator {
+    /// Creates a new instance of IdAllocator that will be used to manage the
+    /// allocation and release of ids from the interval specified by
+    /// `range_base` and `range_max`
+    pub fn new(range_base: u32, range_max: u32) -> Self {
+        IdAllocator {
+            range_base,
+            next_id: range_base,
+            range_end: range_max,
+            freed_ids: BTreeSet::<u32>::new(),
+        }
+    }
+
+    /// Checks if a specified id belongs to the range that this instance of
+    /// `IdAllocator` is managing and if it was not freed before.
+    fn check_id(&self, id: u32) -> Result {
```

Comment:
nit: Can we rename this to `id_in_range` instead? That would make the comment unnecessary and it speaks more to what the function is doing.